lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -44,29 +53,57 @@ void CTreeShapFeatureImportance::shap(const TRowRef& row, TShapWriter writer) {
         return;
     }
 
+    using TTreeShapVec = std::vector<std::function<void(const TTree&)>>;
+
     auto encodedRow{m_Encoder->encode(row)};
-    m_ShapValues.assign(m_Encoder->numberInputColumns(),
-                        las::zero((*m_Forest)[0][0].value()));
-    for (const auto& tree : *m_Forest) {
-        this->shapRecursive(tree, encodedRow, 0, 1.0, 1.0, -1,
-                            CSplitPath{m_PathStorage.begin(), m_ScaleStorage.begin()},
-                            0, m_ShapValues);
+
+    if (m_PerThreadShapValues.size() == 1) {
+        m_ReducedShapValues.assign(m_Encoder->numberInputColumns(),
+                                   las::zero((*m_Forest)[0][0].value()));
+        for (const auto& tree : *m_Forest) {
+            this->shapRecursive(
+                tree, encodedRow, 0, 1.0, 1.0, -1,
+                CSplitPath{m_PathStorage[0].begin(), m_ScaleStorage[0].begin()},
+                0, m_ReducedShapValues);
+        }
+    } else {
+        TTreeShapVec computeTreeShap;
+        computeTreeShap.reserve(m_PerThreadShapValues.size());
+        for (std::size_t i = 0; i < m_PerThreadShapValues.size(); ++i) {
+            m_PerThreadShapValues[i].assign(m_Encoder->numberInputColumns(),
+                                            las::zero((*m_Forest)[0][0].value()));
+            computeTreeShap.push_back([&encodedRow, i, this](const TTree& tree) {
+                this->shapRecursive(tree, encodedRow, 0, 1.0, 1.0, -1,
+                                    CSplitPath{m_PathStorage[i].begin(),
+                                               m_ScaleStorage[i].begin()},
+                                    0, m_PerThreadShapValues[i]);
+            });
+        }
+
+        core::parallel_for_each(m_Forest->begin(), m_Forest->size(), computeTreeShap);
```

Comment:
Will this work if you use `begin()` and `end()` instead of `size()`?

Reply:
This calls [this](https://github.com/elastic/ml-cpp/pull/1277/files#diff-75e42b5a7634c55d254341641a79d6fcR340) overload. I used this signature because it matched how the loop was written which I needed to factor out. It does look slightly odd. I'll consider if I can migrate to a version which takes start and end iterators here.

Reply:
Ok I reviewing this code I realised I was missing error checking and handling of reentry for the new functions I exposed. I ended up moving the versions I had into `concurrency_detail` and calling these from the new overloads. A nice side effect is I could could change the signature to take two iterators. See [ea0317c](https://github.com/elastic/ml-cpp/pull/1277/commits/ea0317c1b8eeedd820530f3e3ed2e14c280bb7a7).